src/main/java/org/opensearch/neuralsearch/plugin/query/NeuralQueryBuilder.java

```diff
@@ -172,10 +193,32 @@ private static void parseQueryParams(XContentParser parser, NeuralQueryBuilder n
         }
     }
 
+    @Override
+    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) {
+        if (vectorSupplier() != null && vectorSupplier.get() != null) {
+            return new KNNQueryBuilder(fieldName(), vectorSupplier.get(), k());
+        }
+
+        if (vectorSupplier() == null) {
+            SetOnce<float[]> vectorSetOnce = new SetOnce<>();
+            queryRewriteContext.registerAsyncAction(
+                ((client, actionListener) -> ML_CLIENT.inferenceSentence(modelId(), queryText(), ActionListener.wrap(floatList -> {
+                    vectorSetOnce.set(vectorAsListToArray(floatList));
+                    actionListener.onResponse(null);
+                }, actionListener::onFailure)))
+            );
+            vectorSupplier(vectorSetOnce::get);
+        }
+
+        // Rewrites will continuously happen until the supplier is set and the vector is generated. Rewrites will stop
+        // once this object is returned. Hence, if we get here, we need to return a new object
+        return new NeuralQueryBuilder(fieldName(), queryText(), modelId(), k(), vectorSupplier());
```

Comment:
does this ensure that main thread which called the rewriteQuery comes back and check if the vector response has come or not?

Reply:
Yes thats my understanding. Let me provide better comment on this.

Reply:
Here is what I was basing this off of: https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/query/QueryBuilder.java#L90-L98

Reply:
Here is the logic: https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/query/Rewriteable.java#L83